sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcRelation.scala

```diff
@@ -0,0 +1,176 @@
+      val typeInfo: TypeInfo =
+        TypeInfoUtils.getTypeInfoFromTypeString(orcSchema)
+      standardOI = TypeInfoUtils
+        .getStandardJavaObjectInspectorFromTypeInfo(typeInfo)
+        .asInstanceOf[StructObjectInspector]
+      fieldOIs = standardOI
+        .getAllStructFieldRefs.map(_.getFieldObjectInspector).toArray
+      wrappers = fieldOIs.map(HadoopTypeConverter.wrappers)
+      recordWriter = {
+        outputFormat.getRecordWriter(fs,
+          conf.asInstanceOf[JobConf],
+          file.toUri.getPath, Reporter.NULL)
+          .asInstanceOf[org.apache.hadoop.mapred.RecordWriter[NullWritable, Writable]]
+      }
+    }
+  }
+  override def write(row: Row): Unit = {
+    initWriter()
+    var i = 0
+    val outputData = new Array[Any](fieldOIs.length)
+    while (i < row.length) {
+      outputData(i) = wrappers(i)(row(i))
+      i += 1
+    }
+    val writable = serializer.serialize(outputData, standardOI)
+    recordWriter.write(NullWritable.get(), writable)
+  }
+
+  override def close(): Unit = {
+    if (recordWriter != null) {
+      recordWriter.close(Reporter.NULL)
+    }
+  }
+}
+
+
+@DeveloperApi
+private[sql] case class OrcRelation(override val paths: Array[String],
+    parameters: Map[String, String],
+    maybeSchema: Option[StructType] = None,
+    maybePartitionSpec: Option[PartitionSpec] = None)(
+    @transient val sqlContext: SQLContext)
+  extends FSBasedRelation(paths, maybePartitionSpec)
+  with Logging {
+  self: Product =>
+  @transient val conf = sqlContext.sparkContext.hadoopConfiguration
+
+
+  override def dataSchema: StructType =
+    maybeSchema.getOrElse(OrcFileOperator.readSchema(paths(0), Some(conf)))
+
+  override def outputWriterClass: Class[_ <: OutputWriter] = classOf[OrcOutputWriter]
+  /** Attributes */
+  var output: Seq[Attribute] = schema.toAttributes
+
+  override def needConversion: Boolean = false
+
+  // Equals must also take into account the output attributes so that we can distinguish between
+  // different instances of the same relation,
+  override def equals(other: Any): Boolean = other match {
+    case that: OrcRelation =>
+      paths.toSet == that.paths.toSet &&
+        dataSchema == that.dataSchema &&
+        schema == that.schema &&
+        partitionColumns == that.partitionColumns
+    case _ => false
+  }
+
+  override def hashCode(): Int = {
+    Objects.hashCode(
+      paths.toSet,
+      dataSchema,
+      schema,
+      maybePartitionSpec)
+  }
+  override def buildScan(requiredColumns: Array[String],
+      filters: Array[Filter],
+      inputPaths: Array[String]): RDD[Row] = {
+    val output = StructType(requiredColumns.map(dataSchema(_))).toAttributes
+    OrcTableScan(output, this, filters, inputPaths).execute()
+  }
+}
+
+private[sql] object OrcRelation extends Logging {
+  // Default partition name to use when the partition column value is null or empty string.
+  val DEFAULT_PARTITION_NAME = "__HIVE_DEFAULT_PARTITION__"
```

Comment:
Hm, this is actually the default value of `ConfVar.DEFAULTPARTITIONNAME`. I duplicated it because the partitioning support framework doesn't depend on Hive. You may use the definition from Hive since ORC data source is under Hive.